template/{% if has_ai_rules %}AGENTS.md{% endif %}.jinja

```diff
@@ -0,0 +1,537 @@
+- Follow semver: MAJOR.MINOR.PATCH
+
+**❌ Don't**: Skip version bumps or use non-semantic versions{% endif %}
+
+### Development Approach
+
+**✅ Do**: Start simple and iterate
+- Begin with minimal functionality (e.g., a single command or widget){% if kind == 'frontend-and-server' %}
+- Add backend endpoints only when frontend needs them{% endif %}
+- Test in running JupyterLab frequently
+- Check browser console{% if kind == 'frontend-and-server' %} and terminal logs{% endif %} for errors
+
+**❌ Don't**: Build complex features without incremental testing
+
+## Common Pitfalls
+
+### Package Management
+
+**✅ Do**: Use `jlpm` consistently (JupyterLab's pinned Yarn)
+```bash
+jlpm install
+jlpm build
+```
+
+**❌ Don't**: Mix package managers or lockfiles
+- Don't use both `yarn.lock` and `package-lock.json`
+- Don't run `npm install` if you started with `jlpm`
+
+### Path Handling
+
+{% if kind == 'frontend-and-server' %}**✅ Do**: Use proper path utilities in `{{ python_name }}/handlers.py`
+```python
+from jupyter_server.utils import url_path_join
+route = url_path_join(base_url, "{{ python_name }}", "endpoint")
+```
+
+**❌ Don't**: Hardcode paths like `"/{{ python_name }}/endpoint"`
+
+{% endif %}**✅ Do**: Use relative imports in TypeScript (`src/` files)
+```typescript
+import { MyWidget } from './widgets/MyWidget';
+```
+
+**❌ Don't**: Use absolute paths or assume specific directory structures
+
+### Error Handling
+
+**✅ Do**: Wrap async operations in try-catch (in `src/api.ts`, widget code)
+```typescript
+try {
+  const data = await fetchData();
+} catch (err) {
+  showErrorMessage('Failed to fetch data');
+}
+```
+
+**❌ Don't**: Let errors propagate silently or crash the extension
+
+### CSS and Styling
+
+**✅ Do**: Namespace all CSS in `style/index.css`
+```css
+.{{ python_name | replace('_', '-') }}-widget {
+  padding: 8px;
+}
+```
+
+**❌ Don't**: Use generic class names like `.widget` or `.button`
+
+### Resource Cleanup
+
+**✅ Do**: Dispose resources in widget `dispose()` methods
+```typescript
+dispose(): void {
+  this._signal.disconnect();
+  super.dispose();
+}
+```
+
+**❌ Don't**: Leave event listeners or signal connections active after disposal
+
+{% if kind == 'frontend-and-server' %}### Backend Integration
+
+**✅ Do**: Use `APIHandler` for CORS support in `{{ python_name }}/handlers.py`
+```python
+from jupyter_server.base.handlers import APIHandler
+
+class MyHandler(APIHandler):
+    pass
+```
+
+**❌ Don't**: Use plain `tornado.web.RequestHandler`
+
+**✅ Do**: Use relative imports within your package
+```python
+from .handlers import setup_handlers
+```
+
+**❌ Don't**: Use absolute imports like `from {{ python_name }}.handlers import ...`
+{% endif %}
+
+## Quick Reference
+
+### Key Identifiers
+
+Use these patterns consistently throughout your code:
+
+- **Plugin ID** (in `src/index.ts`): `'{{ python_name }}:plugin'`
+- **Command IDs** (in `src/index.ts`): `'{{ python_name }}:command-name'`
```

Comment:
```suggestion
- **Command IDs** (in `src/commands.ts`): `'{{ python_name }}:command-name'`
```

With many commands we may want to extract command registry information to constants in another file, let's make this consistent with advice elsewhere in the file.